diff --git a/config.def.h b/config.def.h
index defa426..9cb55bb 100644
--- a/config.def.h
+++ b/config.def.h
@@ -1,11 +1,11 @@
 /* See LICENSE file for copyright and license details. */
 
 /* appearance */
-static const char font[]        = "monospace:size=9";
-static const char* normbgcolor  = "#222222";
-static const char* normfgcolor  = "#cccccc";
-static const char* selbgcolor   = "#555555";
-static const char* selfgcolor   = "#ffffff";
+static const char font[]        = "IosevkaTerm Nerd Font:size=10";
+static const char* normbgcolor  = "#cccccc";
+static const char* normfgcolor  = "#333333";
+static const char* selbgcolor   = "#ffffff";
+static const char* selfgcolor   = "#000000";
 static const char* urgbgcolor   = "#111111";
 static const char* urgfgcolor   = "#cc0000";
 static const char before[]      = "<";
