diff --git a/config.def.h b/config.def.h
index 0dffcd7..ea5b2fa 100644
--- a/config.def.h
+++ b/config.def.h
@@ -175,9 +175,9 @@ static uint forcemousemod = ShiftMask;
  * Beware that overloading Button1 will disable the selection.
  */
 static MouseShortcut mshortcuts[] = {
-	/* mask                 button   function        argument       release */
-	{ XK_ANY_MOD,           Button4, kscrollup,      {.i = 1} },
-	{ XK_ANY_MOD,           Button5, kscrolldown,    {.i = 1} },
+	/* mask                 button   function        argument       release  altscrn*/
+	{ XK_ANY_MOD,           Button4, kscrollup,      {.i = 1},      0,       -1 },
+	{ XK_ANY_MOD,           Button5, kscrolldown,    {.i = 1},      0,       -1 },
 	{ XK_ANY_MOD,           Button2, selpaste,       {.i = 0},      1 },
 	{ ShiftMask,            Button4, ttysend,        {.s = "\033[5;2~"} },
 	{ XK_ANY_MOD,           Button4, ttysend,        {.s = "\031"} },
diff --git a/st.c b/st.c
index 6e76caf..057f81d 100644
--- a/st.c
+++ b/st.c
@@ -1182,6 +1182,12 @@ tloaddefscreen(int clear, int loadcursor)
 	}
 }
 
+int
+tisaltscreen(void)
+{
+	return IS_SET(MODE_ALTSCREEN);
+}
+
 void
 tloadaltscreen(int clear, int savecursor)
 {
diff --git a/st.h b/st.h
index 8c40157..11f9cc4 100644
--- a/st.h
+++ b/st.h
@@ -93,6 +93,7 @@ void toggleprinter(const Arg *);
 int tattrset(int);
 void tnew(int, int);
 void tresize(int, int);
+int tisaltscreen(void);
 void tsetdirtattr(int);
 void ttyhangup(void);
 int ttynew(const char *, char *, const char *, char **);
diff --git a/x.c b/x.c
index 0da7644..18428f3 100644
--- a/x.c
+++ b/x.c
@@ -35,6 +35,7 @@ typedef struct {
 	void (*func)(const Arg *);
 	const Arg arg;
 	uint  release;
+	int  altscrn;  /* 0: don't care, -1: not alt screen, 1: alt screen */
 } MouseShortcut;
 
 typedef struct {
@@ -456,6 +457,7 @@ mouseaction(XEvent *e, uint release)
 	for (ms = mshortcuts; ms < mshortcuts + LEN(mshortcuts); ms++) {
 		if (ms->release == release &&
 		    ms->button == e->xbutton.button &&
+		    (!ms->altscrn || (ms->altscrn == (tisaltscreen() ? 1 : -1))) &&
 		    (match(ms->mod, state) ||  /* exact or forced */
 		     match(ms->mod, state & ~forcemousemod))) {
 			ms->func(&(ms->arg));
