diff --git a/config.def.h b/config.def.h
index 91ab8ca..8517b4e 100644
--- a/config.def.h
+++ b/config.def.h
@@ -189,6 +189,10 @@ static MouseShortcut mshortcuts[] = {
 
 static Shortcut shortcuts[] = {
 	/* mask                 keysym          function        argument */
+	{ ControlMask,          XK_equal,       zoom,           {.f = +2} },
+	{ ControlMask,          XK_minus,       zoom,           {.f = -2} },
+	{ ControlMask,          XK_0,           zoomreset,      {.f =  0} },
+	{ ShiftMask,            XK_BackSpace,   ttysend,        {.s = "\177"} },
 	{ XK_ANY_MOD,           XK_Break,       sendbreak,      {.i =  0} },
 	{ ControlMask,          XK_Print,       toggleprinter,  {.i =  0} },
 	{ ShiftMask,            XK_Print,       printscreen,    {.i =  0} },
